source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -373,6 +373,26 @@ void AuthenticatorImpl::doneWithStatus(const Status& status) {
   ENVOY_LOG(debug, "{}: JWT token verification completed with: {}", name(),
             ::google::jwt_verify::getStatusString(status));
 
+  if (Status::Ok != status) {
+    // Forward the failed status to dynamic metadata
+    ENVOY_LOG(info, "status is: {}", ::google::jwt_verify::getStatusString(status));
+
+    if (!jwks_cache_.findByProvider(*provider_)
```

Comment:
Just check if `jwks_data_` is null, if no, uses it directly.   `jwks_data_` is null case is rare,   you could not map the JWT token to any `provider` in that case.   Not handling that case is fine.


```
if (!jwks_data_  && 
  !jwks_data_->getJwtProvider().failed_status_in_metadata().empty()) {
```

Reply:
@danieldradware and I saw that in the case of a missing JWT ("Status::JwtMissed") the jwks_data_ is NULL and so this failure reason is missed and can't be output in the metadata, so inorder to handle it as well we used this condition instead.

Reply:
Normally,  in the config, your `requirement` will have `provider_name`.  authenticator will have a valid `provider_`,   so `jwks_data_` is not null. 

If you use `allow_missing`  or `allow_missing_or_fail` as a `requirement`,  then `provider_` in the authenticator is null,  we need to use `iss` field from the token to find `provider`,  if that is not a known issuer,  then we get `UnknownIssuer` status.   Only in this case, `jwks_data_` is NULL.    We could not associate a token with any `provider`.

Your change doesn't work either,  as `provider_` is empty in the latter case. 

Current issue is:  all our config are specified under `provider`  and we support multiple `provider`.  We need to find out which one we should use when we work on a token.

But in most case,  there is only one `provider` in the whole config.    So maybe we can introduce a concept of `getDefault()`  in `JwksCache` class.   If there is only one `provider`,  `getDefault()` will return it, otherwise, it will fail.

With that your code will be:

```
const std::string failed_status_in_metadata = 
  (jwks_data_) ?   jwks_data_->getJwtProvider().failed_status_in_metadata() 
 \:   jwks_cache_.getDefault()->getJwtProvider().failed_status_in_metadata() 
   
```